shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -175,18 +175,26 @@ - (void)ensureSemanticsEnabled {
 
 - (void)setViewController:(FlutterViewController*)viewController {
   FML_DCHECK(self.iosPlatformView);
-  _viewController = [viewController getWeakPtr];
+  _viewController = viewController ? [viewController getWeakPtr] : fml::WeakPtr<FlutterViewController>();
```

Comment:
This LGTM, but what was happening previously? Calling getWeakPtr on nil when we're unsetting it actually produces a WeakPtr instance which crashes on .get()?

@dnfield might have some opinions too.

Reply:
viewController is nil with type FlutterViewController. We call the WeakPtrFactory on a undefined reference which try to wrap itself. I guess it just some garbage, and observation will break the compiler for some reason which prevent me from observe what is inside.

Reply:
schrodinger's cat

Reply:
Hm.  Sending a message to `nil` results in `nil`. But we expect the weak ptr to be a reference.  I think this is ok.

Reply:
When I tried viewController.get(), it does try to grab what is inside except it fails the thread check. I am not familiar with objective c. If viewController is nil, will viewController.get() still work?

Reply:
Errrr failing the thread check is not good.  That's a different bug.  What thread are you trying to call this from?  It shoul dbe the platform thread.

/cc @gw280 @chinmaygarde

Reply:
it is a driver test calls this https://github.com/flutter/engine/blob/8ed7a2c21f989d382d64059484330edfe36046cf/testing/scenario_app/ios/Scenarios/ScenariosTests/AppLifecycleTests.m#L291
and it only happens in my local machine. I believe it is passing luci because somehow the platform thread shared the same thread as driver test?

Reply:
Can we check (via xcode debugger or something) whether the remaining issue is indeed a threading issue? If so, let's use the right threads in the test via something like https://github.com/flutter/engine/blob/master/fml/synchronization/count_down_latch_unittests.cc#L24

Reply:
We end up doing something similar on Android too when running instrumentation tests (which drives from a different thread while the engine expects main thread access). https://github.com/flutter/engine/blob/master/testing/scenario_app/android/app/src/androidTest/java/dev/flutter/scenarios/EngineLaunchE2ETest.java

Reply:
I think the code here is correct. The weakptr is created in test thread and unwrapped in the same thread. It is because the nil vc will cause some issue.
consider this case
```
FlutterViewController* testvc = nil;
fml::WeakPtr<FlutterViewController> test = [testvc getWeakPtr];
```
I believed [testvc getWeakPtr] will return a static_cast<fml::WeakPtr<FlutterViewController>> version of nil. this will result in a unintialized weakptr whose creating thread is 0x0;
When you later do test.get(), it tries to compare creating thread 0x0 with current thread and failed.

Reply:
either way, I think this is tangential to this pr.